0000-ilp-ping-protocol/0000-ilp-ping-protocol.md

```diff
@@ -0,0 +1,78 @@
+
+All Interledger implementations _SHOULD_ respond to ping protocol requests, unless a node has a specific reason not to, such as security or privacy concerns.
+
+## Overview
+This RFC defines two ping protocol modes: unidirectional and bidirectional.
+
+* **Unidirectional**: A ping Initiator sends a specially formed ILP Prepare packet, and the Receiver either fulfills or rejects the packet according to the rules defined in [Unidirectional Mode](#Unidirectional Mode).
+* **Bidirectional**: An Initiator sends a specially formed "ping" packet that prompts the Receiver to send an additional "pong" packet. Upon receiving this pong packet, the Initiator either fulfills or rejects according to the rules defined in [Bidirectional Mode](#Bidirectional Mode), thus prompting the Receiver to fulfill or reject the original ping packet.
+
+The primary difference between the two modes is that unidirectional mode allows an Initiator to test connectivity from the Initiator to the Recipient only, whereas bidirectional mode allows the Initiator to test connectivity in both directions.
+
+## Terminology
+
+* The **Initiator** is the Interledger node initiating the ping request.
+* The **Recipient** is the Interledger node responding to the Initiator's request.
+* **Source amount** is the amount debited from the sender of an ILP payment.
+* **Destination amount** is the amount credited to the Recipient of an ILP payment.
+* The **Known Preimage** is the [ASCII](https://tools.ietf.org/html/rfc20) string `pingpingpingpingpingpingpingping`, which in binary is `70696E6770696E6770696E6770696E6770696E6770696E6770696E6770696E67`. This value is used in Unidirectional mode only. 
+* The **Known Condition** is the SHA256 hash of the Known Preimage, which when encoded using Base64 is `jAC8DGFPZPfh4AtZpXuvXFe2oRmpDVSvSJg2oT+bx34=`. This value is used in Unidirectional mode only.
+
+## Unidirectional Mode
+
+1. Initiator sends a ILP Prepare packet using the following details:
+   - **Destination**: Recipient's Interledger address.
+   - **Amount**: Any amount chosen by the Initiator.
+   - **Expiry**: Any appropriate expiry.
+   - **Condition**: The Known Condition, as defined in [Terminology](#terminology).
+   
+2. Upon receiving the Prepare packet, the Recipient MUST fulfill the payment using the following information:
+   - **Fulfillment**: The Known Fulfillment, as defined above.
+
+Note that a Recipient _MAY_ reject the payment if appropriate, for example due to an insufficient amount, invalid expiry, or for other factors.
+
+## Bidirectional Mode
+
+1. Initiator generates a new random 32-byte value to use as a Fulfilment (`F`) and generates the corresponding condition (`C`).
+2. Initiator sends an ILP Prepare packet to the Recipient with the following details:
+   - **Destination**: The Recipient's Interledger address.
+   - **Amount**: Any amount chosen by the Initiator.
```

Comment:
Can you describe how the initiator would choose the amount? Should it be zero? Should it be non-zero? Are they going to get that money back?

Reply:
I'm not sure I have a single answer here. In general I prefer the idea of picking something > zero here, because I think a Connector should generally reject all 0-value packets (this is Interledger after all, and it should be trivial to send _some_ amount of value).

I'm not sure what the value should be though for pinging. Intuitively, it seems like the rule should be "high enough to make it through the path, but as low as possible."

Anything more specific come to mind?

Reply:
> because I think a Connector should generally reject all 0-amount packets (this is Interledger after all, and it should be trivial to send some amount of value)

I very much disagree with this. Zero amount packets are used in STREAM to send data, and I think there is no point in disallowing them. If you disallow it, I would just send an unfulfillable packet of a non-zero amount to achieve the same effect while unnecessarily tying up liquidity. You can't stop people from sending packets just to transmit data so you might as well embrace it and just rate limit the number of packets sent for any reason if that is a concern.